community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/CountsMigrator.java

```diff
@@ -0,0 +1,175 @@
+
+import static org.neo4j.kernel.impl.store.MetaDataStore.DEFAULT_NAME;
+import static org.neo4j.kernel.impl.store.format.RecordFormatSelector.selectForVersion;
+import static org.neo4j.kernel.impl.storemigration.FileOperation.DELETE;
+import static org.neo4j.kernel.impl.storemigration.FileOperation.MOVE;
+
+/**
+ * Rebuilds the count store during migration.
+ * <p>
+ * Since the database may or may not reside in the upgrade directory, depending on whether the new format has
+ * different capabilities or not, we rebuild the count store using the information the store directory if we fail to
+ * open the store in the upgrade directory.
+ * <p>
+ * Just one out of many potential participants in a {@link StoreUpgrader migration}.
+ *
+ * @see StoreUpgrader
+ */
+public class CountsMigrator extends AbstractStoreMigrationParticipant
+{
+
+    private static final Iterable<StoreFile> COUNTS_STORE_FILES = Iterables
+            .iterable( StoreFile.COUNTS_STORE_LEFT, StoreFile.COUNTS_STORE_RIGHT );
+    private final Config config;
+    private final LogService logService;
+    private final FileSystemAbstraction fileSystem;
+    private final PageCache pageCache;
+
+    public CountsMigrator( FileSystemAbstraction fileSystem, PageCache pageCache, Config config, LogService logService )
+    {
+        super( "Count rebuilding" );
+        this.fileSystem = fileSystem;
+        this.pageCache = pageCache;
+        this.config = config;
+        this.logService = logService;
+    }
+
+    @Override
+    public void migrate( File storeDir, File migrationDir, MigrationProgressMonitor.Section progressMonitor,
+            String versionToMigrateFrom, String versionToMigrateTo ) throws IOException
+    {
+        if ( countStoreRebuildRequired( versionToMigrateFrom ) )
+        {
+            // create counters from scratch
+            StoreFile.fileOperation( DELETE, fileSystem, migrationDir, migrationDir, COUNTS_STORE_FILES, true, null,
+                    StoreFileType.STORE );
+            File neoStore = new File( storeDir, DEFAULT_NAME );
+            long lastTxId = MetaDataStore.getRecord( pageCache, neoStore, Position.LAST_TRANSACTION_ID );
+            try
+            {
+                rebuildCountsFromScratch( migrationDir, migrationDir, lastTxId, progressMonitor, versionToMigrateTo,
+                        pageCache, NullLogProvider.getInstance() );
+            }
+            catch ( StoreFailureException e )
+            {
+                //This means that we did not perform a full migration, as the formats had the same capabilities. Thus
+                // we should use the store directory for information when rebuilding the count store. Note that we
+                // still put the new count store in the migration directory.
+                rebuildCountsFromScratch( storeDir, migrationDir, lastTxId, progressMonitor, versionToMigrateFrom,
+                        pageCache, logService.getInternalLogProvider() );
```

Comment:
this is current migration output of migration from 3.1 to 3.3 in debug log:
```
2017-07-05 07:54:48.051+0000 INFO [o.n.k.i.s.p.StoreMigrator] Starting upgrade of database
2017-07-05 07:54:48.085+0000 INFO [o.n.k.i.s.p.StoreMigrator] Migrating Indexes (1/4):
2017-07-05 07:54:48.089+0000 INFO [o.n.k.i.s.p.StoreMigrator]   10% completed
2017-07-05 07:54:48.089+0000 INFO [o.n.k.i.s.p.StoreMigrator]   20% completed
2017-07-05 07:54:48.090+0000 INFO [o.n.k.i.s.p.StoreMigrator]   30% completed
2017-07-05 07:54:48.090+0000 INFO [o.n.k.i.s.p.StoreMigrator]   40% completed
2017-07-05 07:54:48.090+0000 INFO [o.n.k.i.s.p.StoreMigrator]   50% completed
2017-07-05 07:54:48.090+0000 INFO [o.n.k.i.s.p.StoreMigrator]   60% completed
2017-07-05 07:54:48.090+0000 INFO [o.n.k.i.s.p.StoreMigrator]   70% completed
2017-07-05 07:54:48.090+0000 INFO [o.n.k.i.s.p.StoreMigrator]   80% completed
2017-07-05 07:54:48.090+0000 INFO [o.n.k.i.s.p.StoreMigrator]   90% completed
2017-07-05 07:54:48.090+0000 INFO [o.n.k.i.s.p.StoreMigrator]   100% completed
2017-07-05 07:54:48.090+0000 INFO [o.n.k.i.s.p.StoreMigrator] Migrating Legacy indexes (2/4):
2017-07-05 07:54:48.091+0000 INFO [o.n.k.i.s.p.StoreMigrator]   10% completed
2017-07-05 07:54:48.091+0000 INFO [o.n.k.i.s.p.StoreMigrator]   20% completed
2017-07-05 07:54:48.091+0000 INFO [o.n.k.i.s.p.StoreMigrator]   30% completed
2017-07-05 07:54:48.091+0000 INFO [o.n.k.i.s.p.StoreMigrator]   40% completed
2017-07-05 07:54:48.091+0000 INFO [o.n.k.i.s.p.StoreMigrator]   50% completed
2017-07-05 07:54:48.092+0000 INFO [o.n.k.i.s.p.StoreMigrator]   60% completed
2017-07-05 07:54:48.092+0000 INFO [o.n.k.i.s.p.StoreMigrator]   70% completed
2017-07-05 07:54:48.092+0000 INFO [o.n.k.i.s.p.StoreMigrator]   80% completed
2017-07-05 07:54:48.092+0000 INFO [o.n.k.i.s.p.StoreMigrator]   90% completed
2017-07-05 07:54:48.092+0000 INFO [o.n.k.i.s.p.StoreMigrator]   100% completed
2017-07-05 07:54:48.092+0000 INFO [o.n.k.i.s.p.StoreMigrator] Migrating Store files (3/4):
2017-07-05 07:54:48.108+0000 INFO [o.n.k.i.s.p.StoreMigrator]   10% completed
2017-07-05 07:54:48.108+0000 INFO [o.n.k.i.s.p.StoreMigrator]   20% completed
2017-07-05 07:54:48.108+0000 INFO [o.n.k.i.s.p.StoreMigrator]   30% completed
2017-07-05 07:54:48.108+0000 INFO [o.n.k.i.s.p.StoreMigrator]   40% completed
2017-07-05 07:54:48.108+0000 INFO [o.n.k.i.s.p.StoreMigrator]   50% completed
2017-07-05 07:54:48.108+0000 INFO [o.n.k.i.s.p.StoreMigrator]   60% completed
2017-07-05 07:54:48.108+0000 INFO [o.n.k.i.s.p.StoreMigrator]   70% completed
2017-07-05 07:54:48.109+0000 INFO [o.n.k.i.s.p.StoreMigrator]   80% completed
2017-07-05 07:54:48.109+0000 INFO [o.n.k.i.s.p.StoreMigrator]   90% completed
2017-07-05 07:54:48.109+0000 INFO [o.n.k.i.s.p.StoreMigrator]   100% completed
2017-07-05 07:54:48.109+0000 INFO [o.n.k.i.s.p.StoreMigrator] Migrating Count rebuilding (4/4):
2017-07-05 07:56:47.371+0000 INFO [o.n.k.i.s.p.StoreMigrator]   10% completed
2017-07-05 07:56:47.388+0000 INFO [o.n.k.i.s.p.StoreMigrator]   20% completed
2017-07-05 07:56:47.403+0000 INFO [o.n.k.i.s.p.StoreMigrator]   30% completed
2017-07-05 07:56:47.416+0000 INFO [o.n.k.i.s.p.StoreMigrator]   40% completed
2017-07-05 07:56:47.430+0000 INFO [o.n.k.i.s.p.StoreMigrator]   50% completed
2017-07-05 07:56:47.444+0000 INFO [o.n.k.i.s.p.StoreMigrator]   60% completed
2017-07-05 07:56:47.459+0000 INFO [o.n.k.i.s.p.StoreMigrator]   70% completed
2017-07-05 07:56:47.501+0000 INFO [o.n.k.i.s.p.StoreMigrator]   80% completed
2017-07-05 07:56:47.612+0000 INFO [o.n.k.i.s.p.StoreMigrator]   90% completed
2017-07-05 07:56:47.613+0000 INFO [o.n.k.i.s.p.StoreMigrator]   100% completed
2017-07-05 07:57:44.313+0000 INFO [o.n.k.i.s.p.StoreMigrator] Successfully finished upgrade of database
2017-07-05 07:57:44.328+0000 INFO [o.n.k.i.s.f.RecordFormatSelector] Selected RecordFormat:StandardV3_2[v0.A.8] record format from store data/databases/graph.db
2017-07-05 07:57:44.328+0000 INFO [o.n.k.i.s.f.RecordFormatSelector] Format not configured. Selected format from the store: RecordFormat:StandardV3_2[v0.A.8]
2017-07-05 07:57:44.362+0000 ERROR [o.n.k.i.s.c.CountsTracker] Failed to open counts store file: data/databases/graph.db/neostore.counts.db.b Format header/trailer has changed. Expected format:`['N', 'e', 'o', 'C', 'o', 'u', 'n', 't', 'S', 't', 'o', 'r', 'e', 0x00, 0x02, 'V']`, actual:`['N', 'e', 'o', 'C', 'o', 'u', 'n', 't', 'S', 't', 'o', 'r', 'e', 0x00, 0x01, 'V']`.
java.lang.IllegalStateException: Format header/trailer has changed. Expected format:`['N', 'e', 'o', 'C', 'o', 'u', 'n', 't', 'S', 't', 'o', 'r', 'e', 0x00, 0x02, 'V']`, actual:`['N', 'e', 'o', 'C', 'o', 'u', 'n', 't', 'S', 't', 'o', 'r', 'e', 0x00, 0x01, 'V']`.
	at org.neo4j.kernel.impl.store.kvstore.MetadataCollector$State.readFormatSpecifier(MetadataCollector.java:242)
	at org.neo4j.kernel.impl.store.kvstore.MetadataCollector$State.access$100(MetadataCollector.java:114)
	at org.neo4j.kernel.impl.store.kvstore.MetadataCollector$State$1.visit(MetadataCollector.java:121)
	at org.neo4j.kernel.impl.store.kvstore.MetadataCollector.visit(MetadataCollector.java:69)
	at org.neo4j.kernel.impl.store.kvstore.MetadataCollector.visit(MetadataCollector.java:29)
	at org.neo4j.kernel.impl.store.kvstore.KeyValueStoreFile.visitKeyValuePairs(KeyValueStoreFile.java:192)
	at org.neo4j.kernel.impl.store.kvstore.KeyValueStoreFile.scanAll(KeyValueStoreFile.java:183)
	at org.neo4j.kernel.impl.store.kvstore.KeyValueStoreFileFormat.open(KeyValueStoreFileFormat.java:287)
	at org.neo4j.kernel.impl.store.kvstore.KeyValueStoreFileFormat.openStore(KeyValueStoreFileFormat.java:78)
	at org.neo4j.kernel.impl.store.kvstore.RotationStrategy.open(RotationStrategy.java:65)
	at org.neo4j.kernel.impl.store.kvstore.DeadState$Stopped.initialize(DeadState.java:125)
	at org.neo4j.kernel.impl.store.kvstore.AbstractKeyValueStore.init(AbstractKeyValueStore.java:172)
	at org.neo4j.kernel.impl.store.NeoStores.createCountStore(NeoStores.java:640)
	at org.neo4j.kernel.impl.store.StoreType$15.open(StoreType.java:156)
	at org.neo4j.kernel.impl.store.StoreType$15.open(StoreType.java:151)
	at org.neo4j.kernel.impl.store.NeoStores.openStore(NeoStores.java:255)
	at org.neo4j.kernel.impl.store.NeoStores.getOrCreateStore(NeoStores.java:300)
	at org.neo4j.kernel.impl.store.NeoStores.<init>(NeoStores.java:144)
	at org.neo4j.kernel.impl.store.StoreFactory.openNeoStores(StoreFactory.java:169)
	at org.neo4j.kernel.impl.store.StoreFactory.openAllNeoStores(StoreFactory.java:133)
	at org.neo4j.kernel.impl.storageengine.impl.recordstorage.RecordStorageEngine.<init>(RecordStorageEngine.java:214)
	at org.neo4j.kernel.NeoStoreDataSource.buildStorageEngine(NeoStoreDataSource.java:583)
	at org.neo4j.kernel.NeoStoreDataSource.start(NeoStoreDataSource.java:441)
	at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:442)
	at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:107)
	at org.neo4j.kernel.impl.transaction.state.DataSourceManager.start(DataSourceManager.java:100)
	at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:442)
	at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:107)
	at org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory.initFacade(GraphDatabaseFacadeFactory.java:205)
	at org.neo4j.graphdb.EnterpriseGraphDatabase.<init>(EnterpriseGraphDatabase.java:44)
	at org.neo4j.server.enterprise.EnterpriseNeoServer.lambda$static$1(EnterpriseNeoServer.java:74)
	at org.neo4j.server.database.LifecycleManagingDatabase.start(LifecycleManagingDatabase.java:89)
	at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:442)
	at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:107)
	at org.neo4j.server.AbstractNeoServer.start(AbstractNeoServer.java:207)
	at org.neo4j.server.ServerBootstrapper.start(ServerBootstrapper.java:107)
	at org.neo4j.server.ServerBootstrapper.start(ServerBootstrapper.java:80)
	at org.neo4j.server.enterprise.EnterpriseEntryPoint.main(EnterpriseEntryPoint.java:36)
```
that looks wrong